### PR TITLE
Add 'filesystem' view

### LINKED ISSUE #78: Filesystem-like view of the game tree
Optional command group consisting of commands for navigation and modification of the game tree.

evaera@DataModel `cd Workspace`
evaera@Workspace `ls` 
evaera@Workspace `prop Baseplate CanCollide false`
evaera@Workspace `rm Baseplate`

Potential commands `cd`, `pwd`, `rm`, `cp`, `mv`, `make <instance> [name]`, `prop <name> <property> [value to set]`

`prop`'s value parameter could use dynamic Type from #93

### PR DESCRIPTION
**Pull request by Reshiram110**
This adds a 'filesystem' like view to Cmdr, allowing the developer to navigate the gametree with bash-like commands such as `cd`, `ls`, and more.

Closes #78.

### CODE REVIEW

File: Cmdr/BuiltInCommands/Tree/cd.lua [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		--[[ Changing active instance ]]--
+		if treeView.View == "Client" then
+			local Instance;
+			
+			if string.split(stringPath,".")[1] == "game" then -- Absolute path specified
+				Instance = util.GetInstanceFromStringPath(stringPath)
+			else -- Relative path specified
+				local StartingInstance;
+				if not game:IsAncestorOf(treeView.WorkingInstance) and treeView.WorkingInstance ~= game then --Instance is in a detached state from the datamodel
+					context:Reply("The current working instance has been destroyed, unexpected behavior may occur!\n")
```
Comment (evaera) [inline Cmdr/BuiltInCommands/Tree/cd.lua]: Detached state doesn't necessarily mean destroyed; for example, character models are removed when the player dies but they do not get destroyed. I would change message to:

```suggestion
					context:Reply("Warning: the current working instance has been detached from the Data Model\n")
```

File: Cmdr/BuiltInCommands/Tree/cd.lua [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+			else
+
+				--[[ Updating view when/if instance is reparented/destroyed ]]--
+				if AncestryChanged~=nil then
+					AncestryChanged:Disconnect()
+				end
+				AncestryChanged = Instance.AncestryChanged:connect(function(_,Parent)
+					context.Cmdr:SetPrompt(("%s.%s:%s"):format(game.Name, treeView.View, util.GetInstanceFullName(Instance)))
+					if not game:IsAncestorOf(Instance) and Instance ~= game then
+						context:Reply("The current working instance has been destroyed, unexpected behavior may occur!\n")
```
Comment (evaera) [inline Cmdr/BuiltInCommands/Tree/cd.lua]: Move this error message into a local constant at top of file

File: Cmdr/BuiltInCommands/Tree/cd.lua [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			if Instance == nil then
+				return ("'%s' is not recognized as a valid Instance."):format(stringPath)
+			else
+
+				--[[ Updating view when/if instance is reparented/destroyed ]]--
+				if AncestryChanged~=nil then
+					AncestryChanged:Disconnect()
+				end
+				AncestryChanged = Instance.AncestryChanged:connect(function(_,Parent)
+					context.Cmdr:SetPrompt(("%s.%s:%s"):format(game.Name, treeView.View, util.GetInstanceFullName(Instance)))
```
Comment (evaera) [inline Cmdr/BuiltInCommands/Tree/cd.lua]: Swap into `Client/Server.game.etc` after extracting logic into module

File: Cmdr/BuiltInCommands/Tree/ls.lua [hunk 4]
```diff
@@ -0,0 +1,54 @@
+return {
+	Name = "ls",
+	Description = "Lists all of the children of the current working instance.",
+	Group = "DefaultTree",
+	Args = {
+		{
+			Type = "string",
+			Name = "Instance",
+			Description = "The instance to list the children of. Can be a relative path or afull path.",
```
Comment (evaera) [inline Cmdr/BuiltInCommands/Tree/ls.lua]:
```suggestion
			Description = "The instance to list the children of. Can be a relative path or a full path.",
```

File: Cmdr/CmdrClient/init.lua [hunk 5]
```diff
@@ -49,12 +49,18 @@ function Cmdr:SetActivationKeys (keysArray)
...
---- Sets the place name label on the interface
-function Cmdr:SetPlaceName (name)
-	self.PlaceName = name
+--- Sets the prompt on the interface
+function Cmdr:SetPrompt (prompt)
+	self.Prompt = prompt
 	Interface.Window:UpdateLabel()
 end
 
+--- Sets the place name label on the interface
```
Comment (evaera) [inline Cmdr/CmdrClient/init.lua]: Replace with 

```lua
Cmdr.SetPlaceName = Cmdr.SetPrompt
```

File: Cmdr/Shared/Util.lua [hunk 6]
```diff
@@ -465,4 +465,59 @@ function Util.EmulateTabstops(text, tabWidth)
...
+
+	if stringPath == "game" then
+		return game
+	elseif currentObjectReference == game then
+		stringPath = stringPath:gsub("^game%.", "")
+	else
+		stringPath = stringPath:gsub("^"..startingReference.Name.."%.", "")
+	end
+
+	for _, objectName in pairs(Util.SplitStringSimple(stringPath, ".")) do
```
Comment (evaera) [inline Cmdr/Shared/Util.lua]:
```suggestion
	for _, objectName in pairs(string.split(stringPath, ".")) do
```

File: Cmdr/Shared/Util.lua [hunk 7]
```diff
@@ -465,4 +465,59 @@ function Util.EmulateTabstops(text, tabWidth)
...
+		else
+			return nil
+		end
+	end
+
+	return currentObjectReference
+end
+
+--- Returns the TRUE full pathname of a given instance. (i.e. Workspace will return as "game.Workspace" instead of just "Workspace")
+function Util.GetInstanceFullName(Instance)
```
Comment (evaera) [inline Cmdr/Shared/Util.lua]: This logic is too complicated, just return

```lua
return instance == game and "game" or (game:IsAncestorOf(instance) and "game." or "") .. instance:GetFullName()
```

File: Cmdr/Shared/Util.lua [hunk 8]
```diff
@@ -465,4 +465,59 @@ function Util.EmulateTabstops(text, tabWidth)
...
+		FullName = "game"
+	else
+		FullName = "game."..Instance:GetFullName()
+	end
+
+	return FullName
+end
+
+--- Returns the root ancestor of a given instance.
+function Util.GetInstanceRootAncestor(Instance)
```
Comment (evaera) [inline Cmdr/Shared/Util.lua]: Change to:

```lua

while currentObjectReference.Parent do
  currentObjectReference = currentObjectReference.Parent
end

return currentObjectReference
```

File: Cmdr/BuiltInCommands/Tree/cd.lua (hunk 2 above)
Comment (Reshiram110) [reply Cmdr/BuiltInCommands/Tree/cd.lua]: Fixed in https://github.com/evaera/Cmdr/pull/98/commits/65b56b26fd806038216f8f6da3a8826d9cb75622

File: Cmdr/BuiltInCommands/Tree/cd.lua (hunk 1 above)
Comment (Reshiram110) [reply Cmdr/BuiltInCommands/Tree/cd.lua]: Fixed in https://github.com/evaera/Cmdr/pull/98/commits/93a49f5f0dbb977604498dd57f86aed92381deb0 and https://github.com/evaera/Cmdr/pull/98/commits/65b56b26fd806038216f8f6da3a8826d9cb75622

**Comment by evaera** [comment]
What is the status on this pr?

**Comment by Reshiram110** [comment]
Apologies about the late response - I've been incredibly busy and have not had the time to work on it. I should be getting more time sometime within the next week or so.

File: Cmdr/CmdrClient/init.lua (hunk 5 above)
Comment (Reshiram110) [reply Cmdr/CmdrClient/init.lua:58]: Fixed in https://github.com/evaera/Cmdr/pull/98/commits/18f76d59c5a9d4c79b14c1845a9c2591b4313f8f

**Comment by Reshiram110** [comment]
Heya, noticed that you closed this PR - would you want someone else to work on the PR? I don't really have the time to work on it unfortunately, I have a lot on my plate.

(Tried DMing you on discord about it, but it seems you've disabled DMs and/or blocked me)

**Pull request closed without merging**